### PR TITLE
Object.assign() preserves simplicity of its target

### PR DESCRIPTION
**Pull request by gaearon**
Alternative to https://github.com/facebook/prepack/pull/1450, per the discussion there.

### CODE REVIEW

File: src/intrinsics/ecma262/Object.js [hunk 1]
```diff
@@ -61,6 +61,15 @@ export default function(realm: Realm): NativeFunctionValue {
     let to = To.ToObjectPartial(realm, target);
     let to_must_be_partial = false;
 
+    if (to.isSimpleObject()) {
+      // isSimpleObject() is lazily calculated based on properties for normal objects.
+      // However we might later make it partial in this function. In that case,
+      // the default will be that the object is *not* simple. But we know that
+      // it was simple before, and that Object.assign() can't possibly copy any
+      // getters or setters onto it. Therefore, it is safe to claim it stays simple.
+      to.makeSimple();
```
Comment (sebmarkbage) [inline src/intrinsics/ecma262/Object.js]: This makes this object permanently simple even if we do:

```js
var x = Object.assign({}, partial);
var y = 0;
Object.defineProperty(x, 'foo', { get() { y++ } });
Object.assign({}, x);
result = y;
```

We need that defineProperty to unmark this as simple.

Comment (gaearon) [reply src/intrinsics/ecma262/Object.js]: That's what I thought but I asked @hermanventer and he said

>The idea of a simple object is simply that it's safe to read and write unknown properties. If there is a getter of setter involved that is OK as long as there is no observable side effects. **Of course if the property is known and we have the definition of the getter or setter then that is OK too.**

and

>Demoting a simple object is not a supported scenario. It is probably not a good idea.

So I'm confused over this. I'm also not sure what assumptions Prepack codebase makes around "simple" (i.e. does it treat it as "has no getters/setters" or "has no unknown getters/setters") in practice, regardless of the original intention.

For example would this be safe if there were known getters/setters:

https://github.com/facebook/prepack/blob/6203b6b125e916815fac848d06ca86ed7f73d4dc/src/evaluators/BinaryExpression.js#L133

?

This

https://github.com/facebook/prepack/blob/6203b6b125e916815fac848d06ca86ed7f73d4dc/src/methods/to.js#L783

?

I'm not sure. The fact that it's used in so few places also makes it hard for me to tell what it's supposed to represent in practice, and if it's consistent today or not.

Comment (sebmarkbage) [reply src/intrinsics/ecma262/Object.js]: If that's the case, then it seems to me that we should only set this explicit flag if we end up in the unknown partial case.

**Comment by trueadm** [comment]
Can we add the tests in the other PR too please?

**Comment by gaearon** [comment]
Yep, I will.

**Comment by gaearon** [comment]
I found a weird case that this PR "fixes". I don't understand why.

```js
var React = require('react');
this['React'] = React;

function App(props) {
  var copyOfProps = {};
  var copyOfProps2 = {};
  Object.assign(copyOfProps, props, {x: 20});
  Object.assign(copyOfProps2, copyOfProps);
  return copyOfProps2.x;
}

App.getTrials = function(renderer, Root) {
  renderer.update(<Root x={10} />);
  return [['simple render with object assign', renderer.toJSON()]];
};

if (this.__registerReactComponentRoot) {
  __registerReactComponentRoot(App);
}

module.exports = App;
```

Without this PR, this gives different output (`10` vs `20`). This is worrying me.

**Comment by trueadm** [comment]
https://github.com/facebook/prepack/blob/master/src/intrinsics/ecma262/Object.js#L71

We should be using `const` not `let` for the loop, as we're creating a closure in the loop and it will always be of the last value.

**Comment by sebmarkbage** [comment]
Hm. That is worrying.

File: src/intrinsics/ecma262/Object.js [hunk 2]
```diff
@@ -140,7 +140,16 @@ export default function(realm: Realm): NativeFunctionValue {
     }
 
     // 5. Return to.
-    if (to_must_be_partial) to.makePartial();
+    if (to_must_be_partial) {
+      const to_was_simple = to.isSimpleObject();
```
Comment (hermanventer) [inline src/intrinsics/ecma262/Object.js]: simple and partial should be orthogonal, so don't make the simplicity of to subject to it being partial.

**Comment by gaearon** [comment]
@hermanventer

Can you please provide more details? The reason I need this fix is they’re already not quite orthogonal.

`isSimpleObject()` returns true for a known object (because it enumerates the properties). But as soon as I make it partial, `isSimpleObject()` starts returning false (because now it looks at the flag). So changing one already changes the other.

In my case I need to make an object partial without un-making it simple (if it was already simple). The crux of the issue is that simplicity is calculated differently for partial and non-partial object.

**Comment by gaearon** [comment]
@hermanventer Is this change sufficient?

**Comment by hermanventer** [comment]
You are right. I misunderstood your original change. The resulting object from Object assign is going to be simple whenever it is partial because of checks done earlier on. Although objects can be any combination of simple and partial, the simple flag is only needed if an object is partial because in that case it is not possible to compute simplicity. That makes the flags not quite orthogonal. It also means that your original formulation is better than your most recent change.

To make it easier for future readers to figure this out, would you please assert that the to_must_be_partial necessarily implies that to.isSimpleObject() is true and refine the comment a bit to make it clearer that explicitly setting the flag is necessary because it the to object is being marked as partial.

File: src/intrinsics/ecma262/Object.js [hunk 3]
```diff
@@ -141,14 +141,17 @@ export default function(realm: Realm): NativeFunctionValue {
 
     // 5. Return to.
     if (to_must_be_partial) {
-      const to_was_simple = to.isSimpleObject();
+      // We allow partial sources (and make `to` partial) only if `to` has no keys.
```
Comment (hermanventer) [inline src/intrinsics/ecma262/Object.js]: And the sources are simple.

File: src/intrinsics/ecma262/Object.js [hunk 4]
```diff
@@ -141,14 +141,17 @@ export default function(realm: Realm): NativeFunctionValue {
...
+
       to.makePartial();
 
-      if (to_was_simple) {
-        // Making it partial has switched isSimple from calculated to a flag.
-        // Set that flag explicitly if this was simple before.
-        to.makeSimple();
-      }
+      // Partial objects (and `to` is now partial) aren't considered simple by
+      // default because we can't iterate over their properties.
```
Comment (hermanventer) [inline src/intrinsics/ecma262/Object.js]: over all of their

File: src/intrinsics/ecma262/Object.js [hunk 5]
```diff
@@ -141,14 +141,17 @@ export default function(realm: Realm): NativeFunctionValue {
...
+      invariant(to.isSimpleObject());
+
       to.makePartial();
 
-      if (to_was_simple) {
-        // Making it partial has switched isSimple from calculated to a flag.
-        // Set that flag explicitly if this was simple before.
-        to.makeSimple();
-      }
+      // Partial objects (and `to` is now partial) aren't considered simple by
```
Comment (hermanventer) [inline src/intrinsics/ecma262/Object.js]: Partial objects (and `to` is now partial) can't be calculated to be simple

File: src/intrinsics/ecma262/Object.js [hunk 6]
```diff
@@ -141,14 +141,17 @@ export default function(realm: Realm): NativeFunctionValue {
...
+      // default because we can't iterate over their properties.
+      // However we already asserted above that `to` is simple.
```
Comment (hermanventer) [inline src/intrinsics/ecma262/Object.js]: We already established above that `to` is simple, so

File: src/intrinsics/ecma262/Object.js [hunk 7]
```diff
@@ -141,14 +141,17 @@ export default function(realm: Realm): NativeFunctionValue {
...
+      // Set the flag explicitly so `to` doesn't lose its simple status.
```
Comment (hermanventer) [inline src/intrinsics/ecma262/Object.js]: set the _isSimple flag

**Pull request closed without merging**